src/EFCache/CachingReader.cs

```diff
@@ -137,7 +137,16 @@ public override short GetInt16(int ordinal)
 
         public override int GetInt32(int ordinal)
         {
-            return (int)GetValue(ordinal);
+            var value = GetValue(ordinal);
+            switch (value)
+            {
+                case int intValue:
+                    return intValue;
+                case long longValue:
+                    return Convert.ToInt32(longValue);
+                default:
+                    throw new NotImplementedException($"{value.GetType()}");
+            }
```

Comment:
Any reason why we cannot do:

```suggestion
            return Convert.ToInt32(GetValue(ordinal));
```